File: backend/src/main/java/com/project/controller/UserController.java
package com.project.controller;

import java.util.List;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.project.dto.user.request.PasswordUpdateDTO;
import com.project.dto.user.request.UserUpdateDTO;
import com.project.dto.user.response.ProfessionalViewDTO;
import com.project.entities.User;
import com.project.exception.ApiResponse;
import com.project.exception.ResponseBuilder;
import com.project.security.CustomUserDetails;
import com.project.service.UserService;

import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;

@RestController
@RequestMapping("/api/user")
@RequiredArgsConstructor
public class UserController {

    private final UserService userService;

    /* ================= PROFILE ================= */

    @PutMapping("/profile")
    public ResponseEntity<ApiResponse<User>> updateProfile(
            @RequestBody @Valid UserUpdateDTO dto,
            Authentication auth) {

        CustomUserDetails user =
                (CustomUserDetails) auth.getPrincipal();

        User updated =
                userService.updateProfile(user.getUserId(), dto);

        return ResponseBuilder.success(
                "Profile updated successfully",
                null,
                HttpStatus.OK
        );
    }
    
    
    @GetMapping("/profile")
    public ResponseEntity<ApiResponse<UserUpdateDTO>> updateProfile(
            Authentication auth) {

        CustomUserDetails user =
                (CustomUserDetails) auth.getPrincipal();

        UserUpdateDTO profile =
                userService.getProfile(user.getUserId());

        return ResponseBuilder.success(
                "Profile found",
                profile,
                HttpStatus.OK
        );
    }

    /* ================= PASSWORD ================= */

    @PutMapping("/password")
    public ResponseEntity<ApiResponse<Object>> updatePassword(
            @RequestBody @Valid PasswordUpdateDTO dto,
            Authentication auth) {

        CustomUserDetails user =
                (CustomUserDetails) auth.getPrincipal();

        userService.updatePassword(user.getUserId(), dto);

        return ResponseBuilder.success(
                "Password updated successfully",
                null,
                HttpStatus.NO_CONTENT
        );
    }

    /* ================= APPOINTMENT ================= */

    @PutMapping("/appointments/{id}/cancel")
    public ResponseEntity<ApiResponse<Object>> cancelAppointment(
            @PathVariable Long id,
            Authentication auth) {

        CustomUserDetails user =
                (CustomUserDetails) auth.getPrincipal();

        userService.cancelConfirmedAppointment(
                id, user.getUserId());

        return ResponseBuilder.success(
                "Appointment cancelled successfully",
                null,
                HttpStatus.NO_CONTENT
        );
    }

    /* ================= PROFESSIONALS ================= */

    @GetMapping("/professionals")
    public ResponseEntity<ApiResponse<List<ProfessionalViewDTO>>> getProfessionals() {

        List<ProfessionalViewDTO> professionals =
                userService.getAllProfessionals();

        return ResponseBuilder.success(
                "Professionals fetched successfully",
                professionals,
                HttpStatus.OK
        );
    }
}

File: backend_dotnet/SafeMind.API/Controllers/UserController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using SafeMind.Core.DTOs.User.Request;
using SafeMind.Core.DTOs.User.Response;
using SafeMind.Core.Entities;
using SafeMind.Core.Interfaces;
using System.Security.Claims;
using SafeMind.Core.Responses;

namespace SafeMind.API.Controllers
{
    [ApiController]
    [Route("api/user")]
    [Authorize]
    public class UserController : ControllerBase
    {
        private readonly IUserService _userService;

        public UserController(IUserService userService)
        {
            _userService = userService;
        }

        private long GetCurrentUserId()
        {
            var userIdClaim = User.FindFirst("userId");
            if (userIdClaim == null) throw new UnauthorizedAccessException("Invalid token");
            return long.Parse(userIdClaim.Value);
        }

        /* ================= PROFILE ================= */

        [HttpPut("profile")]
        public async Task<IActionResult> UpdateProfile([FromBody] UserUpdateDTO dto)
        {
            try
            {
                var userId = GetCurrentUserId();
                var updatedUser = await _userService.UpdateProfileAsync(userId, dto);
                
                return Ok(ApiResponse<object>.SuccessResponse("Profile updated successfully", null));
            }
            catch (Exception ex)
            {
                return BadRequest(ApiResponse<object>.FailureResponse(ex.Message));
            }
        }

        [HttpGet("profile")]
        public async Task<IActionResult> GetProfile()
        {
            try
            {
                var userId = GetCurrentUserId();
                var profile = await _userService.GetProfileAsync(userId);
                
                return Ok(ApiResponse<UserUpdateDTO>.SuccessResponse("Profile found", profile));
            }
            catch (Exception ex)
            {
                 return BadRequest(ApiResponse<object>.FailureResponse(ex.Message));
            }
        }

        /* ================= PASSWORD ================= */

        [HttpPut("password")]
        public async Task<IActionResult> UpdatePassword([FromBody] PasswordUpdateDTO dto)
        {
             try
            {
                var userId = GetCurrentUserId();
                await _userService.UpdatePasswordAsync(userId, dto);
                
                return Ok(ApiResponse<object>.SuccessResponse("Password updated successfully", null));
            }
            catch (Exception ex)
            {
                 return BadRequest(ApiResponse<object>.FailureResponse(ex.Message));
            }
        }

        /* ================= APPOINTMENT ================= */

        [HttpPut("appointments/{id}/cancel")]
        public async Task<IActionResult> CancelAppointment(long id)
        {
             try
            {
                var userId = GetCurrentUserId();
                await _userService.CancelConfirmedAppointmentAsync(id, userId);
                
                return Ok(ApiResponse<object>.SuccessResponse("Appointment cancelled successfully", null));
            }
             catch (Exception ex)
            {
                 return BadRequest(ApiResponse<object>.FailureResponse(ex.Message));
            }
        }

        /* ================= PROFESSIONALS ================= */

        [HttpGet("professionals")]
        [AllowAnonymous] // Assuming listing professionals might be public? Java Controller didn't specify, likely authenticated.
        // Actually Java code: @RequestMapping("/api/user") -> likely for logged in users.
        // But `getAllProfessionals` might be used by patients.
        // Let's keep [Authorize] from class level, unless requested otherwise.
        public async Task<IActionResult> GetProfessionals()
        {
             try
            {
                var professionals = await _userService.GetAllProfessionalsAsync();
                
                 return Ok(ApiResponse<IEnumerable<ProfessionalViewDTO>>.SuccessResponse("Professionals fetched successfully", professionals));
            }
             catch (Exception ex)
            {
                 return BadRequest(ApiResponse<object>.FailureResponse(ex.Message));
            }
        }
    }
}
